### PR TITLE
🌱 Update conditions.Set function to set LastTransitionTime only when status of condition changes

### LINKED ISSUE #11033: conditions Utility to let specify the entire condition
### What would you like to be added (User Story)?

A new conditions utility to upsert conditions with the given LastTransistionTime and state fields.

### Detailed Description

The LastTransistionTime definition Says

```
Last time the condition transitioned from one status to another.
This should be when the underlying condition changed. If that is not known, then using the time when
the API field changed is acceptable.
```

Would be better to have a utility in conditions to let it specify the entire condition and upsert the condition when the condition type already exists. The Utility should add / update the condition with the given state fields which include

- Type
- Status
- Reason
- Severity
- Message

And the LastTransistionTime specified in the new condition to add.

### Anything else you would like to add?

The Utility may be used to track the transition of the state in any cloud provider / any other resources.

### Label(s) to be applied

/kind feature
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

### PR DESCRIPTION
**Pull request by Karthik-K-N**
**What this PR does / why we need it**:

This PR existing behavior of Set condition function as follows

Existing behavior

1. Update condition.LastTransitionTime to current time whenever any one of the fields(Status, Reason, Severity, Message) changes in new condition.
2. Sets the condition.LastTransitionTime to newConidtion.LastTransitionTime if defined only when the newCondition does not exist.

Proposed behavior

1.  Update condition.LastTransitionTime to current time only when Status field of new condition differ from existing condition status.
2. Sets the condition.LastTransitionTime to newCondition.LastTransitionTime  if defined whether the newCondition already exist or new.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #11033

### CODE REVIEW

**Comment by Karthik-K-N** [comment]
@fabriziopandini @sbueringer Please review when you get time. Thank you.

**Comment by chrischdi** [comment]
/retitle 🌱 Update conditions.Set function to set LastTransitionTime only when status of condition changes

File: util/conditions/setter_test.go [hunk 1]
```diff
@@ -180,6 +184,22 @@ func TestSetLastTransitionTime(t *testing.T) {
...
+			to:   setterWithConditions(fooWithLastTransitionTime),
+			new:  fooWithAnotherStateWithLastTransitionTime,
+			LastTransitionTimeCheck: func(g *WithT, lastTransitionTime metav1.Time) {
+				g.Expect(lastTransitionTime).To(Equal(y))
+			},
+		},
+		{
+			name: "Set a condition that already exists but with different Message should preserve the last transition time",
+			to:   setterWithConditions(fooWithLastTransitionTime),
+			new:  fooWithBarMessage,
```
Comment (fabriziopandini) [inline util/conditions/setter_test.go:198]: What about adding also a test where the new condition has a last transition time time set (but it gets ignored)

Comment (Karthik-K-N) [reply util/conditions/setter_test.go:217]: Added a new test in separate commit, Please let me know if anything else needs to be added. Thank you.

**Comment by Karthik-K-N** [comment]
Please take a look at this PR, when you get chance. Thank you.

**Comment by sbueringer** [comment]
Heyho,
sorry should have brought this up earlier. The more I'm thinking about this the more I'm wondering if we should really make this change. Basically we are now moving towards v1beta2 conditions and this util will be removed sooner or later. So I'm not sure if we really should make a behavioral change. The util for v1beta2 already behaves the correct way.

@fabriziopandini @vincepri WDYT?

If this is really absolutely needed, I would recommend that we an additional Set func with a different name and keep the current Set exactly as is.

**Comment by fabriziopandini** [comment]
I'm ok in adding a separated func (e.g. SetEx), it is a practical weys to avoid issues on old utils considering the deprecation/removal plan

**Comment by Karthik-K-N** [comment]
I can do that, But who will be the caller of new function, is it like if the providers are interested they can make use of it right?

**Comment by sbueringer** [comment]
Yup providers and the folks that opened and engaged on the issue. But nothing in this repo we are moving to v1beta2 conditions instead and the new v1beta2 utils already have this behavior

**Comment by Karthik-K-N** [comment]
Created a new function with the requested logic, Please take a look and suggest better name if any. Thanks

**Comment by Karthik-K-N** [comment]
> Perfect, thank you very much!
> 
> /lgtm /approve

Thanks, I forgot to change the PR title, It may mislead.

**Comment by sbueringer** [comment]
You can still edit the title now. The release notes tool will use the latest PR title (merge commit won't but it's okay, that happens)

**Pull request merged**